gtsfm/ui/process_graph_generator.py

```diff
@@ -0,0 +1,160 @@
+                continue
+
+            # get UI metadata of class
+            metadata = cls_type.get_ui_metadata()
+
+            # skip duplicates
+            # happens when concrete classes implement an abstract class without overwriting get_ui_metadata()
+            if metadata in unique_metadata:
+                continue
+            unique_metadata.add(metadata)
+
+            # create an empty plate for each unique parent_plate name
+            plate = metadata.parent_plate
+            if plate not in self._plate_to_cluster:
+                # if no plate, add straight to main graph
+                if plate is None:
+                    continue
+
+                new_cluster = pydot.Cluster(plate, label=plate)
+                self._plate_to_cluster[plate] = new_cluster
+
+        return unique_metadata
+
+    def _add_nodes_to_graph(self, metadata: UiMetadata) -> None:
+        """Add connected blue/gray nodes to relevant subgraph based on UiMetadata.
+
+        Also, auto-casts metadata.input_products and metadata.output_products to tuples of strings.
+
+        Why? A common error is to define a one-element tuple like so:
+          > input_products = ("Input Product")
+        but this is actually parsed as a raw string. The correct usage is:
+          > input_products = ("Input Product",)
+        which is unintuitive. Auto-cast here prevents unexpected behavior for developers.
+
+        Args:
+            metadata: UiMetadata object to add nodes/edges for
+        """
+
+        display_name = metadata.display_name
+
+        # autocast strings to one-element tuples
+        output_products = metadata.output_products
+        if type(output_products) == str:
+            output_products = (output_products,)
+
+        # autocast strings to one-element tuples
+        input_products = metadata.input_products
+        if type(input_products) == str:
+            input_products = (input_products,)
+
+        cluster = self._main_graph
+        if metadata.parent_plate is not None:
+            cluster = self._plate_to_cluster[metadata.parent_plate]
+
+        # add Nodes for processes and output_products in the same plate
+        # don't add input_products as Nodes to ensure output_products are in the same plate as their processes
+        cluster.add_node(pydot.Node(display_name, shape=NODE_SHAPE, style=NODE_STYLE, fillcolor=PROCESS_FILLCOLOR))
+        for product_name in output_products:
+            cluster.add_node(pydot.Node(product_name, shape=NODE_SHAPE, style=NODE_STYLE, fillcolor=PRODUCT_FILLCOLOR))
+
+        # add process -> output_product edges
+        for output_product_name in output_products:
+            self._main_graph.add_edge(pydot.Edge(display_name, output_product_name, color=ARROW_COLOR))
+
+        # add input_product -> process edges
+        for input_product_name in input_products:
+            self._main_graph.add_edge(pydot.Edge(input_product_name, display_name, color=ARROW_COLOR))
```

Comment:
Do we detect cases where we have a typo and the input products of one Process does not match the output product of other process?

Reply:
No, though that will be obvious in the graph. In the future, people will likely only be adding a few nodes at a time, so I don't think typos will be a large issue.